src/envoy/auth/config.cc

```diff
@@ -15,199 +15,119 @@
-#include <fstream>
-#include <iostream>
-#include <memory>
-#include <string>
-#include <vector>
 
 namespace Envoy {
 namespace Http {
 namespace Auth {
+namespace {
+
+// Default public key cache cache duration: 5 minutes.
+const int64_t kPubKeyCacheExpirationSec = 600;
+
+}  // namespace
 
-void AsyncClientCallbacks::onSuccess(MessagePtr &&response) {
-  std::string status = response->headers().Status()->value().c_str();
-  if (status == "200") {
-    ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: success",
-              cluster_->name());
-    std::string body;
-    if (response->body()) {
-      auto len = response->body()->length();
-      body = std::string(static_cast<char *>(response->body()->linearize(len)),
-                         len);
-    } else {
-      ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: body is null",
-                cluster_->name());
+std::string IssuerInfo::Validate() const {
+  if (!pubkey_value.empty()) {
+    auto pubkey = Pubkeys::CreateFrom(pubkey_value, pubkey_type);
+    if (pubkey->GetStatus() != Status::OK) {
+      return std::string("Invalid public key value: ") + pubkey_value;
     }
-    cb_(true, body);
   } else {
-    ENVOY_LOG(debug,
-              "AsyncClientCallbacks [cluster = {}]: response status code {}",
-              cluster_->name(), status);
-    cb_(false, "");
+    if (uri == "") {
+      return "Missing public key server uri";
+    }
+    if (cluster == "") {
+      return "Missing public key server cluster";
+    }
   }
-}
-void AsyncClientCallbacks::onFailure(AsyncClient::FailureReason) {
-  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: failed",
-            cluster_->name());
-  cb_(false, "");
+  return "";
 }
 
-void AsyncClientCallbacks::Call(const std::string &uri) {
-  ENVOY_LOG(debug, "AsyncClientCallbacks [cluster = {}]: {} {}",
-            cluster_->name(), __func__, uri);
-  // Example:
-  // uri  = "https://example.com/certs"
-  // pos  :          ^
-  // pos1 :                     ^
-  // host = "example.com"
-  // path = "/certs"
-  auto pos = uri.find("://");
-  pos = pos == std::string::npos ? 0 : pos + 3;  // Start position of host
-  auto pos1 = uri.find("/", pos);
-  if (pos1 == std::string::npos) pos1 = uri.length();
-  std::string host = uri.substr(pos, pos1 - pos);
-  std::string path = "/" + uri.substr(pos1 + 1);
-
-  MessagePtr message(new RequestMessageImpl());
-  message->headers().insertMethod().value().setReference(
-      Http::Headers::get().MethodValues.Get);
-  message->headers().insertPath().value(path);
-  message->headers().insertHost().value(host);
-
-  request_ = cm_.httpAsyncClientForCluster(cluster_->name())
-                 .send(std::move(message), *this, timeout_);
+Config::Config(std::vector<IssuerInfo>&& issuers)
+    : issuers_(std::move(issuers)) {
+  ENVOY_LOG(debug, "Config: {}", __func__);
 }
 
-void AsyncClientCallbacks::Cancel() { request_->cancel(); }
+Config::Config(const Json::Object& config) {
```

Comment:
Consider define a proto instead of JSON, with envoy provided validate rules. JSON conversion will handled by envoy and we don't need handle it manually.

Reply:
You mean FilterFactory allows passing Proto now?   But before we upstream the filter,  how does Envoy know our config proto?

Reply:
https://github.com/envoyproxy/envoy/blob/master/include/envoy/server/filter_config.h#L287